File: py_bst/BST.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

""" 
An implementation of a complete binary search tree. The 'complete' part of the 
this tree comes where each node will always have two children, where one will 
either be a value and the other None or both will be None.
"""


class Leaf(object):

    def __init__(self, value):
        self.value = value
        self.parent = None
        self.left = None
        self.right = None

class Tree(object):

    def __init__(self):
        self.root = None


def get_value():
    while True:
        try:
            value = int(raw_input('Enter value > '))
            return value
        except ValueError:
            print "Enter a number"

def create_leaf():
    value = get_value()
    return Leaf(value)

def insert(tree, leaf):
    while tree:
        if leaf.value < tree.value:
            if not tree.left:
                tree.left = leaf
                leaf.parent = tree
                return
            tree = tree.left
        else:
            if not tree.right:
                tree.right = leaf
                leaf.parent = tree
                return
            tree = tree.right
 
def delete(tree, leaf, value):
    leaf = match_leaf(leaf, value)
    if not leaf:
        print "Value <%d> is not in tree!" % value
        return
    if not leaf.left:
        transplant(tree, leaf, leaf.right)
    elif not leaf.right:
        transplant(tree, leaf, leaf.left)
    else:
        _branch = tree_min(leaf.right)
        if _branch.parent != leaf:  #check if branch is child of deleted leaf
            transplant(tree, _branch, _branch.right) #trans branch.rg ;) 
            _branch.right = leaf.right 
            _branch.right.parent = _branch
        transplant(tree, leaf, _branch)
        _branch.left = leaf.left
        _branch.left.parent = _branch

def transplant(tree, leaf, branch):
    if not leaf.parent:
        tree.root = branch
    elif leaf == leaf.parent.left:
        leaf.parent.left = branch
    else:
        leaf.parent.right = branch
    if branch:
        branch.parent = leaf.parent

def match_leaf(branch, value):
    if not branch:
        return branch
    if branch.value > value:
        return match_leaf(branch.left, value)
    elif branch.value < value:
        return match_leaf(branch.right, value)
    return branch
 
def search(tree, value):
    if not tree:
        print "Value <%d> not in tree!" % value
        return
    elif tree.value == value:
        print "Value <%d> is in tree!" % value
        return
    while tree:
        if tree.value == value:
            print "Value <%d> is in tree!" % value
            return
        elif tree.value < value:
            tree = tree.right
        else:
            tree = tree.left
    print "Value <%d> is not in tree!" % value
    return    
    
def tree_walk(tree):
    if tree:
        tree_walk(tree.left)
        print tree.value
        tree_walk(tree.right)

def tree_predecessor(tree):
    if tree.left:
        return tree_max(tree.left)
    parent = tree.parent
    while parent and tree == parent.left:
        tree = parent
        parent = parent.parent
    return parent.value

def tree_successor(tree):
    if tree.right:
        return tree_min(tree.right)
    parent = tree.parent
    while parent and tree == parent.right:
        tree = parent
        parent = parent.parent
    return parent.value

def tree_max(tree):
    while tree.right:
        tree = tree.right
    return tree

def tree_min(tree):
    while tree.left:
        tree = tree.left
    return tree


if __name__ == '__main__':
    tree = Tree() 
    while True:
        print "Select option: [1=insert, 2=delete, 3=search, 4=tree_walk, 5=exit]"
        option = raw_input('Enter number of option > ')
        if option == '1':
            if not tree.root:
                tree.root = create_leaf()
            else:
                leaf = create_leaf()
                insert(tree.root, leaf)     
        elif option == '2':
            value = get_value()
            delete(tree, tree.root, value) 
        elif option == '3':
            value = get_value()
            search(tree.root, value)
        elif option == '4':
            tree_walk(tree.root) 
        else:
            break

File: Java/bst/BST.java
/* Binary search tree */


public class BST 
{
    private static Node root;

    public BST() { root = null; }

    public static class Node
    {
        int value;
        Node parent;
        Node left;
        Node right;

        public Node(int value)
        {
            this.value = value;
            left = null;
            right = null;
            parent = null;
        }
    }

    public static void insert(int value)
    {
        if (root == null) {
            root = new Node(value);
            root.parent = null;
        } else
            insert(root, value);        
    }

    private static void insert(Node node, int value)
    {
        if (node.value < value) {
            if (node.right == null) {
                node.right = new Node(value);
                node.right.parent = node;
            } else
                insert(node.right, value);
        } else {
            if (node.left == null) {
                node.left = new Node(value);
                node.left.parent = node;
            } else
                insert(node.left, value);
        }
    }

    public static void delete(int value)
    {
        if (root == null)
            System.out.printf("Empty\n");
        else
            delete(root, value);
    }

    private static void delete(Node node, int value)
    {
        if (node == null) {
            System.out.printf("Value %d not in tree\n");
            return;
        } else if (node.value > value) {
            delete(node.left, value);
        } else if (node.value < value) {
            delete(node.right, value);
        } else {
            if (node.left == null) {
                transplant(root, node, node.right);
            } else if (node.right == null) {
                transplant(root, node, node.left);
            } else {
                Node successor = min(node.right);
                if (successor != node.right) {
                    transplant(root, successor, successor.right);
                    successor.right = node.right;
                    successor.right.parent = successor;
                }
                transplant(root, node, successor);
                successor.left = node.left;
                node.left.parent = successor;
            }
        }
    }

    private static void transplant(Node root, Node out, Node in)
    {
        if (out.parent == null)
            root = in;
        else if (out.parent.left == out)
            out.parent.left = in;
        else
            out.parent.right = in;
        if (in != null)
            in.parent = out.parent;
    }

    public static void min()
    {
        if (root == null)
            System.out.printf("Empty\n");
        else {
            Node minimum = min(root);
            System.out.printf("Minimum value: %d\n", minimum.value);
        }
    }

    private static Node min(Node node)
    {
        if (node.left != null)
            return min(node.left);
        else
            return node;
    }

    public static void inorder()
    {
        if (root == null)
            System.out.printf("Empty\n");
        else
            inorder(root);
    }

    private static void inorder(Node node)
    {
        if (node == null) return;
        inorder(node.left);
        System.out.printf("Value: %d\n", node.value);
        inorder(node.right);
    }

    public static void preorder()
    {
        if (root == null)
            System.out.printf("Empty\n");
        else
            preorder(root);
    }

    private static void preorder(Node node)
    {
        if (node == null)
            return;

        System.out.printf("Value: %d\n", node.value);
        preorder(node.left);
        preorder(node.right);
    }

    public static void main(String[] args)
    {
        BST tree = new BST();

        int[] tree_values = {20, 43, 1, 39, 99, 71, 14, 55, 83, 4};

        for (int i=0; i < tree_values.length; i++) {
            System.out.printf("Inserting: %d\n", tree_values[i]); 
            insert(tree_values[i]);
        }

        System.out.printf("Inorder print=>\n"); 
        inorder();
        System.out.printf("Preorder print=>\n");
        preorder();

        System.out.printf("Find minimum=>\n");
        min();

        System.out.printf("Deleting %d\n", 43);
        delete(43);
        System.out.printf("Inorder print=>\n"); 
        inorder();
    }
}
